js/src/components/country-names-plus-more.js

```diff
@@ -0,0 +1,44 @@
+/**
+ * External dependencies
+ */
+import { sprintf } from '@wordpress/i18n';
+import { createInterpolateElement } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import useCountryKeyNameMap from '.~/hooks/useCountryKeyNameMap';
+
+const defaultFirstN = 5;
+
+const CountryNamesPlusMore = ( props ) => {
+	const {
+		countries,
+		firstN = defaultFirstN,
+		textWithMore,
+		textWithoutMore,
```

Comment:
> If instead the use sides to specify a prop such as `@param {'rate'|'time'|'minimum-order'} props.kind`, which is mapped inside the component, then there is no need to have the following three additional components.

Deja vu, this sounds oddly familiar with the discussions we had about `AccountCard` and its `appearance` prop. 😆 

I'm not a big fan of the "mapped inside the component", instead I prefer the component composition way. I guess we will just "agree to disagree". 😂 

PS: I also thought about using `CountryNamesPlusMore` component directly without those intermediary "InputControlLabelText" components, but the big bunch of text in the "InputControl" code makes it an eye sore to read the "InputControl" code, so I created these "InputControlLabelText" components to move the big bunch of text out.

Screenshot: left is with using `CountryNamesPlusMore` directly, right is with using "InputControlLabelText".

![image](https://user-images.githubusercontent.com/417342/157461115-c4454d85-3830-4d55-8fd4-05b14d0b77ff.png)

Reply:
I just realized the label text for minimum order was `MinimumOrderInputLabelText` and not consistent with the other two. I have renamed it to `MinimumOrderInputControlLabelText` in [`a068ae2` (#1297)](https://github.com/woocommerce/google-listings-and-ads/pull/1297/commits/a068ae29674794d43601451e7c064796e77c258b).

Reply:
> this sounds oddly familiar with the discussions we had about `AccountCard` and its `appearance` prop.

They may look somewhat similar, but the reasons are different. The primary reason here is: `textWithMore` and `textWithoutMore` are highly correlated with the text transformation logic inside this component. The `%1$s` and `%2$d` format strings and the `<strong>` placeholder are defined inside this component's implementation. The readability and maintainability are relatively low, in my opinion.

And another reason could be considered but I didn't mention is, \<CountryNamesPlusMore\> component is more likely a dedicated component for shipping settings, like those components under .~/components/shipping-rate-section/ directory, etc. It could be designed in a way that is not highly general use. Especially when its flexibility is limited by 18n needed. Building dedicated components at/near the top of application layers is not always a bad design.

> I'm not a big fan of the "mapped inside the component", instead I prefer the component composition way.

Unfortunately, I don't see the obvious cons with this approach. Besides this reason, are there any concerns with this pattern?

> PS: I also thought about using `CountryNamesPlusMore` component directly [...], but the big bunch of text in the "InputControl" code makes it an eye sore to read the "InputControl" code.

Yes, this is not ideal either.

Reply:
> The arguments calling order of `sprintf` and the elements mapping of `createInterpolateElement` are inside this component's implementation, but `%1$s` and `%2$d` format strings and the `<strong>` placeholder are defined outside.

Ahh yes, at first I was concerned and had a bit of doubt about this too. However, after experimenting with it and seeing the outcome, I think it is okay. The high correlation / low readability / low maintainability is between those `***InputControlLabelText` and `CountryNamesPlusMore`. I'm thinking, we can actually move them and group them inside one file or one folder to help ease maintenance. We don't need to expose `CountryNamesPlusMore` component to the whole app.

> Besides this reason, are there any other concerns with this pattern?

If we use the "mapped inside the component" approach, then we would need to expose a prop (e.g. the `@param {'rate'|'time'|'minimum-order'}` like you said earlier) and then use some conditional if or switch statements to render things based on the props. If the number of prop values is small, then it is not a big problem, but if there are many conditional statements with many lines in the file, then it would start to become more difficult, e.g. imagine a dev looking at `<CountryNamesPlusMore type="rate">`, wonder what would be rendered, look into the implementation and see the many lines.

With the component composition approach, we don't need to have the conditional statements, which in my opinion is a plus point. With the above dev scenario, when a dev looks at `<ShippingRateInputControlLabelText>`, wonder what would be rendered, and look into the implementation, there would be fewer lines to go through, because there are no other unrelated things like shipping time and minimum order.

Again, these are personal preferences. 😄 💅 Curious to hear your thoughts too on the above.

Reply:
> I'm thinking, we can actually move them and group them inside one file or one folder to help ease maintenance. We don't need to expose `CountryNamesPlusMore` component to the whole app.

That would be another good way to resolve the concern if move them into CountryNamesPlusMore itself. And even if exposing `CountryNamesPlusMore` component to the whole app is not a problem. For example
```js
// CountryNamesPlusMore
export const minimumOrderTexts = {
	textWithMore: __(
		`Minimum order for <strong>%1$s</strong> + %2$d more`,
		'google-listings-and-ads'
	),
	textWithoutMore: __(
		`Minimum order for <strong>%1$s</strong>`,
		'google-listings-and-ads'
	),
};

// MinimumOrderInputControl

import CountryNamesPlusMore, { minimumOrderTexts } from '.~/components/country-names-plus-more';

return (
  <AppInputPriceControl
    className="gla-minimum-order-input-control"
    label={
      <div className="gla-minimum-order-input-control__label">
        <div>
          <CountryNamesPlusMore
            texts={ minimumOrderTexts }
            countries={ countries }
          />
        </div>
        // Or move <div> into CountryNamesPlusMore
        <div>
          <CountryNamesPlusMore
            texts={ minimumOrderTexts }
            countries={ countries }
          />
        </div>
```

> If we use the "mapped inside the component" approach, then we would need to expose a prop (e.g. the `@param {'rate'|'time'|'minimum-order'}` like you said earlier) and then use some conditional if or switch statements to render things based on the props.

It doesn't need anything complicated or conditional statements, just a `const texts = textsMap[ kind ] || fallback;`. And this is only an example to explain the direction/pattern of decoupling, not necessarily to exactly use this implementation. The above internalization of the highly correlated texts into the CountryNamesPlusMore component is also a way to decouple it.

> If the number of prop values is small, then it is not a big problem, but if there are many conditional statements with many lines in the file, then it would start to become more difficult, e.g. imagine a dev looking at `<CountryNamesPlusMore type="rate">`, wonder what would be rendered, look into the implementation and see the many lines.

My comments were based on that CountryNamesPlusMore could be designed as a dedicated component for shipping settings.

Even if, for some reason, CountryNamesPlusMore requires a lot of conditional statements, it's still better than creating dozens of **\*-input-control-label-text.js** files. And the structure of these large amounts of code is usually simple that even if it has thousands of lines, at least it can be found by using the editor's search function.

```js
const textsMap = {
	minimumOrderTexts: {
		textWithMore: __(
			`Minimum order for <strong>%1$s</strong> + %2$d more`,
			'google-listings-and-ads'
		),
		textWithoutMore: __(
			`Minimum order for <strong>%1$s</strong>`,
			'google-listings-and-ads'
		),
	},
	// other texts in the same structure
};
```

However, if this actually happens, I would look for the feasibility of splitting the "Minimum order for" from the i18n strings and using the annotations provided to translators to compensate for the potentially degraded context of translation.

> With the component composition approach, we don't need to have the conditional statements, which in my opinion is a plus point. With the above dev scenario, when a dev looks at `<ShippingRateInputControlLabelText>`, wonder what would be rendered, and look into the implementation, there would be fewer lines to go through, because there are no other unrelated things like shipping time and minimum order.

Yes, it's easier to read through the implementation with fewer lines. However, doubts can also easily arise in the mind.

As another scenario that actually happened in my first round of code review. When I looked at it, the main thoughts that came to my mind are: Why this extra component is needed? Why are they highly correlated but implemented in a highly coupled manner? Why are there three components with almost duplicate code? What are the benefits of doing this intentionally? Is it to wrap an extra layer of \<div\> or to separate the translation strings? Are there any problems with moving the \<div\> and translation strings into that component?

In addition, after this discussion, a new doubt came to my mind. The fact is that the need for dozens of different translation combinations for this component might not happen now, nor in the known future. So why consider a problem that might not happen and then uses an implementation that reduces readability and maintainability for that non-existent problem now?

And even this implementation doesn't work for situations where a lot of combinations occur. Or would you want to create dozens of **\*-input-control-label-text.js** files for that?

Reply:
> That would be another good way to resolve the concern if move them into CountryNamesPlusMore itself, such as the below example.

I like the `export const minimumOrderTexts` part (i.e. export smaller part of things 🙂 ), however on the consumer side, we have to import two things and use them together, which isn't really my preference. 😅 I mean, it's like "why can't I import one thing and just use that, instead of two things that need to be used together anyway?".

```jsx
          // We import CountryNamesPlusMore and minimumOrderTexts, and use them together.
          <CountryNamesPlusMore 
            texts={ minimumOrderTexts }
            countries={ countries }
          />
```

Maybe there are other ways to achieve that. 

> And the structure of these large amounts of code is usually simple that even if it has thousands of lines

I looked at the `const textsMap` code example and I just want to say I don't prefer the big `textsMap` object. I prefer smaller, specific things. 

> would you want to create dozens of ***-input-control-label-text.js** files for that?

I do prefer dozens of smaller components or js files like that, over (for example) a big `textsMap` object with dozens of lines. 😄